content/zh/blogs/deep-dive-into-the-K8s-request-and-limit.md

```diff
@@ -0,0 +1,256 @@
+
+另外，QOS 作为 K8s 中一种资源保护机制，其主要是针对不可压缩资源比如的内存的一种控制技术，比如在内存中其通过为不同的 pod 和容器构造 OOM 评分，并且通过内核的策略的辅助，从而实现当节点内存资源不足的时候，内核可以按照策略的优先级，优先 kill 掉哪些优先级比较低（分值越高优先级越低）的pod。相关源码如下:
+
+```golang
+func GetContainerOOMScoreAdjust(pod *v1.Pod, container *v1.Container, memoryCapacity int64) int {
+	if types.IsCriticalPod(pod) {
+		// Critical pods should be the last to get killed.
+		return guaranteedOOMScoreAdj
+	}
+
+	switch v1qos.GetPodQOS(pod) {
+	case v1.PodQOSGuaranteed:
+		// Guaranteed containers should be the last to get killed.
+		return guaranteedOOMScoreAdj
+	case v1.PodQOSBestEffort:
+		return besteffortOOMScoreAdj
+	}
+
+	// Burstable containers are a middle tier, between Guaranteed and Best-Effort. Ideally,
+	// we want to protect Burstable containers that consume less memory than requested.
+	// The formula below is a heuristic. A container requesting for 10% of a system's
+	// memory will have an OOM score adjust of 900. If a process in container Y
+	// uses over 10% of memory, its OOM score will be 1000. The idea is that containers
+	// which use more than their request will have an OOM score of 1000 and will be prime
+	// targets for OOM kills.
+	// Note that this is a heuristic, it won't work if a container has many small processes.
+	memoryRequest := container.Resources.Requests.Memory().Value()
+	oomScoreAdjust := 1000 - (1000*memoryRequest)/memoryCapacity
+	// A guaranteed pod using 100% of memory can have an OOM score of 10. Ensure
+	// that burstable pods have a higher OOM score adjustment.
+	if int(oomScoreAdjust) < (1000 + guaranteedOOMScoreAdj) {
+		return (1000 + guaranteedOOMScoreAdj)
+	}
+	// Give burstable pods a higher chance of survival over besteffort pods.
+	if int(oomScoreAdjust) == besteffortOOMScoreAdj {
+		return int(oomScoreAdjust - 1)
+	}
+	return int(oomScoreAdjust)
+}
+```
+
+对于 Guaranteed 级别的 Pod，OOM 参数设置成了 -997 ，对于 Best-Effort 级别的 pod ， OOM 参数设置成了 1000，对于 Burstable 级别的  Pod ， OOM 参数取值从 2 到 999 。因此，如果资源充足，可将 QoS pods 类型均设置为 Guaranteed 。用计算资源换业务性能和稳定性，减少排查问题时间和成本。同时如果想更好的提高资源利用率，业务服务也可以设置为 Guaranteed ，而其他服务根据重要程度可分别设置为 Burstable 或 Best-Effort 。
+
+### 集群的稳定性
+
+集群的稳定性直接决定了其上运行的业务应用的稳定性。而临时性的资源短缺往往是导致集群不稳定的主要因素。集群一旦不稳定，轻则业务应用的性能下降，重则出现相关结点不可用，例如 (CPU 软死锁)[https://unix.stackexchange.com/questions/70377/bug-soft-lockup-cpu-stuck-for-x-seconds]。上面我们已经提到的，通过合理地设置 pod 的 QoS 进一步提高集群稳定性的方法，是集群出现突发的资源短缺的事后应急方案。实际上我们还可以通过(编辑 Kubelet 配置文件)[https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/]来预留一部分系统资源，从而保证当可用计算资源较少时 kubelet 所在节点的稳定性。 这在处理如内存和硬盘之类的不可压缩资源时尤为重要。
+
+## KubeSphere 资源配置实践
+
+前面我们已经深入地探讨了 K8s 中`request`、`limit`这 2 个参数的含义，包括它们是如何间接影响整个集群的状态（包含调度和稳定性等）。实际上这也是 K8s 一直以来的使用痛点，即用户需要对 K8s 集群本身具有一定程度的了解并且具体相关的实战经验，才能保证整个集群的平稳运行。而作为 K8s 的发行版 Kubephere ，极大地降低了 K8s 的学习门槛，配合简介美观的 UI 界面，你会发现有效运维原来是一件如此轻松的事情。下面我们将演示如何在 KubeSphere 平台中配置容器的相关资源。
+
+### 准备工作
+
+您需要创建一个企业空间、一个项目和一个帐户 (project-admin)，务必邀请该帐户到项目中并赋予 admin 角色。有关更多信息，请参见(创建企业空间、项目、帐户和角色)[https://kubesphere.io/zh/docs/quick-start/create-workspace-and-project/]。
```

Comment:
Sorry to weight in here....but perhaps the account to be used here should be `ws-admin`? Have a look at [this](https://github.com/kubesphere/website/pull/831#issuecomment-752885614).

`project-admin` cannot change project quotas once they are set. Here is the [doc](https://kubesphere.io/docs/workspace-administration/project-quotas/).

`project-admin` can set LimitRanges for an individual container. Here is the [doc](https://kubesphere.io/docs/project-administration/container-limit-ranges/).

If you think our doc should be changed, please tell me. Thanks.

Reply:
Yes, you are right. project-admin can only alter LimitRanges once the project was created. And if we want to alter both quotas and Limitranges, we should login as ws-admin. And in this doc I will make a clarification.